### PR TITLE
Make triehash generic over encoding scheme

### PR DESCRIPTION
**Pull request by dvdplm**
Closes https://github.com/paritytech/substrate/issues/664

This makes our Merkle-Patricia Tree ("trie") code generic over both codec and hasher. To do this, it introduces a `TrieStream` trait into `trieroot` crate and a `NodeCodec` trait into `patricia-trie` crate. Neither `trieroot` nor `patricia-trie` are now dependent on RLP. There are two additional crates `patricia-trie-ethereum` (in `test-support` directory) and `substrate-trie`. The latter is not destined for this repo and will be moved to `substrate` after a successful review. The former should probably go into `parity-ethereum` repo but might be handy to have around here for the tests.

Additional changes:
- Rename `patricia_trie` -> `patricia-trie` to maintain convention of `-` over `_` in directory names.
- Refactor `patricia-trie::Node::Branch` to make the child entries `Option`al rather than relying on them being `Null` nodes.
- Rewrote most of trie benchmarking; it now uses Criterion and makes comparisons between RLP and the two Codec alternatives.

Still to do:
- [ ] Remove `println!` logging
- [ ] Remove `substrate-trie`

In case you're wondering, the `parity-codec` based node format is between 1% and 5% faster in random tests, 15% faster in realistic tests and 100% faster (half the time!) for iteration.

### CODE REVIEW

File: triestream/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	fn append_empty_data(&mut self);
+	fn begin_branch(&mut self);
+	fn append_value(&mut self, value: &[u8]);
+	fn append_leaf<H: Hasher>(&mut self, key: &[u8], value: &[u8]) where H: Hasher;
+	fn append_extension(&mut self, key: &[u8]);
+	fn append_substream<H: Hasher>(&mut self, other: Self);
+	fn out(self) -> Vec<u8>;
+	fn as_raw(&self) -> &[u8];
+
+    fn encode(k: &usize) -> Vec<u8>; // Arrgh – `ordered_trie_root` enumerates and rlp-encodes items with `rlp::encode()` so need something similar here. :/
```
Comment (gavofyork) [inline triestream/src/lib.rs]: errant spaces

File: triestream/src/codec_triestream.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+impl TrieStream for CodecTrieStream {
+	fn new() -> Self { Self {buffer: Vec::new() } }
+	fn append_empty_data(&mut self) {
+		self.buffer.push(EMPTY_NODE);
+	}
+
+	fn append_leaf(&mut self, key: &[u8], value: &[u8]) {
+		let mut hpe = hex_prefix_encode_substrate(key, true);
+		self.buffer.push(hpe.next().expect("key is not empty; qed"));
```
Comment (gavofyork) [inline triestream/src/codec_triestream.rs]: should be `self.buffer.extend(key.collect::<Vec<_>>())`

File: triestream/src/codec_triestream.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		println!("[begin_branch] buffer so far: {:#x?}", self.buffer);
+		// TODO: I think this is wrong. I need to know how long the full branch node is I think and
+		// this does not keep track of that.
+	}
+	fn append_value(&mut self, value: &[u8]) {
+		value.encode_to(&mut self.buffer);
+	}
+	fn append_extension(&mut self, key: &[u8]) {
+		let mut hpe = hex_prefix_encode_substrate(key, false);
+		self.buffer.push(hpe.next().expect("key is not empty; qed"));
```
Comment (gavofyork) [inline triestream/src/codec_triestream.rs]: incorrect - shouldn't be here.

replace with `hpe.collect::<Vec<u8>>().encode_to(&mut self.buffer);`

File: patricia_trie/src/parity_node_codec.rs [hunk 4]
```diff
@@ -99,6 +100,19 @@ fn take(input: &mut &[u8], count: usize) -> Option<&[u8]> {
 	Some(r)
 }
 
+fn partial_to_key(partial: &[u8], offset: u8, big: u8) -> Vec<u8> {
+	let nibble_count = partial.len() * 2 + if data[0] & 16 == 16 { 1 } else { 0 };
```
Comment (dvdplm) [inline patricia_trie/src/parity_node_codec.rs]: What is `data`? Missing param?

File: triehash/src/lib.rs [hunk 5]
```diff
@@ -35,37 +38,15 @@ fn shared_prefix_len<T: Eq>(first: &[T], second: &[T]) -> usize {
...
-/// use triehash::ordered_trie_root;
-/// use keccak_hasher::KeccakHasher;
-///
-/// fn main() {
-/// 	let v = &["doe", "reindeer"];
-/// 	let root = "e766d5d51b89dc39d981b41bda63248d7abce4f0225eefd023792a540bcffee3";
-/// 	assert_eq!(ordered_trie_root::<KeccakHasher, _>(v), root.into());
-/// }
-/// ```
-pub fn ordered_trie_root<H, I>(input: I) -> H::Out
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply triehash/src/lib.rs]: that's fair enough - it's pretty trivial.

**Pull request closed without merging**